omniledger/service/contracts.go

```diff
@@ -113,38 +115,112 @@ func LoadDarcFromColl(coll CollectionView, key []byte) (*darc.Darc, error) {
 
 // ContractConfig can only be instantiated once per skipchain, and only for
 // the genesis block.
-func (s *Service) ContractConfig(cdb CollectionView, inst Instruction, coins []Coin) (sc []StateChange, c []Coin, err error) {
+func (s *Service) ContractConfig(cdb CollectionView, scID skipchain.SkipBlockID, inst Instruction, coins []Coin) (sc []StateChange, c []Coin, err error) {
 	if inst.GetType() == SpawnType {
 		return spawnContractConfig(cdb, inst, coins)
 	} else if inst.GetType() == InvokeType {
-		return invokeContractConfig(cdb, inst, coins)
+		return s.invokeContractConfig(cdb, scID, inst, coins)
 	} else {
 		return nil, coins, errors.New("unsupported instruction type")
 	}
 }
 
-func invokeContractConfig(cdb CollectionView, inst Instruction, coins []Coin) (sc []StateChange, c []Coin, err error) {
+func (s *Service) invokeContractConfig(cdb CollectionView, scID skipchain.SkipBlockID, inst Instruction, coins []Coin) (sc []StateChange, c []Coin, err error) {
 	c = coins
-	if inst.Invoke.Command != "update_config" {
+	// There are two situations where we need to change the roster, first
+	// is when it is initiated by the client(s) that holds the genesis
+	// signing key, in thise case we trust the client to do the right
+	// thing. The second is during a view-change, so we need to do
+	// additional validation to make sure a malicious node doesn't freely
+	// change the roster.
+	if inst.Invoke.Command == "update_config" {
```

Comment:
nice catch - there was an extra check outside of the contract that makes sure the new roster has the same length as the old one. it shouldn't be there in the first place because genesis darc holders can modify the config in any way they want.